internal/injector/parse/parser.go

```diff
@@ -0,0 +1,204 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+package parse
+
+import (
+	"errors"
+	"fmt"
+	"go/ast"
+	goparser "go/parser"
+	"go/token"
+	"io"
+	"os"
+	"slices"
+	"sync"
+	"sync/atomic"
+
+	"github.com/DataDog/orchestrion/internal/injector/aspect"
+	"github.com/DataDog/orchestrion/internal/injector/aspect/may"
+)
+
+const maxBytesEagerness = 1 << 19 // 512 KiB
+
+type rawFile struct {
+	name       string
+	mappedName string
+	content    []byte
+}
+
+type File struct {
+	Name    string
+	AstFile *ast.File
+	Aspects []*aspect.Aspect
+}
+
+type Parser struct {
+	fset *token.FileSet // thread-safe data structure
+
+	// rawFiles is an intermediary data structure to store the raw content of the files before parsing them.
+	rawFiles []rawFile
+
+	// filesBytesCount is the sum of the bytes of all files parsed so far.
+	filesBytesCount atomic.Uint64
+
+	// eagerness is a flag that is set to true if at least one file has been parsed and requires all files to be parsed.
+	eagerness atomic.Bool
+
+	// parsedFiles is what is returned by ParseFiles.
+	parsedFiles []File
+
+	wg sync.WaitGroup
+
+	errs   []error
+	errsMu sync.Mutex
+}
+
+func NewParser(fset *token.FileSet, nbFiles int) *Parser {
+	p := &Parser{
+		fset:        fset,
+		rawFiles:    make([]rawFile, nbFiles),
+		parsedFiles: make([]File, nbFiles),
+	}
+
+	p.eagerness.Store(true)
+
+	return p
+}
+
+// ParseFiles return either zero files or all files parsed with their respective aspects
+func (p *Parser) ParseFiles(files []string, aspects []*aspect.Aspect) ([]File, error) {
+	p.wg.Add(len(files))
+
+	for idx, file := range files {
+		fileAspects := make([]*aspect.Aspect, len(aspects))
+		copy(fileAspects, aspects)
+
+		go func(idx int, file string) {
+			defer p.wg.Done()
+			var err error
+			p.rawFiles[idx], err = readFile(file)
+			if err != nil {
+				p.addError(fmt.Errorf("reading %q: %w", file, err))
+				return
+			}
+
+			p.filesBytesCount.Add(uint64(len(p.rawFiles[idx].content)))
+			if p.aspectMayMatch() {
+				fileAspects = p.fileFilterAspects(fileAspects, p.rawFiles[idx])
+				if len(fileAspects) == 0 {
+					// No aspects can match on this file, no need to fill up the File.AstFile field.
+					p.parsedFiles[idx] = File{Name: file}
+					return
+				}
+			}
+
+			p.eagerness.Store(false)
```

Comment:
So we reset `eagerness` to `false` whenever we find the 1st file with aspects... That appears to be the opposite of what I'd have understood from the documentation... The semantics of this is fairly confusing at this stage...